### PR TITLE
fix: missing server ready message

### LINKED ISSUE #42: [server] fix: add `ServerReady` message for new connections

### PR DESCRIPTION
**Pull request by lauti7**
This PR:
- (`server`) sends `ServerReady` message on client connection
- (`transport`) removes the function to determine if it's a `Connect` message or a `Message`
- (`transport`) improves errors for Transports
- (`codegen`) adds constraints to the auto-generated `ClientDefinition` trait for a service
- (`server`) drops `RpcServerPort`s when the transport closes the connection
- (`client`) tries to close the client connection to the server when `RpcClient` drops

Closes #42

### CODE REVIEW

File: rpc/src/client.rs [hunk 1]
```diff
@@ -85,11 +85,6 @@ impl<T: Transport + 'static> RpcClient<T> {
     /// Sends an empty message through the Transport to flag the Transport as "connected"
     async fn establish_connection(transport: T) -> ClientResult<T> {
         // Send empty message to notify connection
-        transport
-            .send(vec![0])
-            .await
-            .map_err(|_| ClientResultError::Client(ClientError::TransportError))?;
-
         match transport.receive().await {
```
Comment (guidota) [inline rpc/src/client.rs:88]: Should we update the comments?

File: rpc/src/transports/web_socket.rs [hunk 2]
```diff
@@ -46,7 +51,9 @@ pub struct WebSocketServer {
 ///
 /// And then attach turn the connection into a transport and attach it to the [`RpcServer`](crate::server::RpcServer)
 ///
-type OnConnectionListener = UnboundedReceiver<Result<Socket, TransportError>>;
+type OnConnectionListener = UnboundedReceiver<Result<Socket, Box<dyn Error + Send + Sync>>>;
+
+pub enum WebSocketServerErrors {}
```
Comment (guidota) [inline rpc/src/transports/web_socket.rs:56]: Is this enum used?

File: rpc/src/transports/web_socket.rs [hunk 3]
```diff
@@ -61,53 +68,61 @@ impl WebSocketServer {
...
+                                if tx_on_connection_listener
+                                    .send(Err(Box::new(error)))
+                                    .is_err()
+                                {
+                                    error!(
+                                            "> WS Server > Error on sending the new ws socket to listener"
+                                    );
+                                    break;
+                                }
+                                error!("> WS Server > Error on upgrading the socket");
```
Comment (guidota) [inline rpc/src/transports/web_socket.rs:114]: Shouldn't this error log happen before sending the error through the channel?

File: rpc/src/server.rs [hunk 4]
```diff
@@ -672,17 +702,24 @@ impl<Context: Send + Sync + 'static, T: Transport + ?Sized + 'static> RpcServer<
...
         message_type: RpcMessageTypes,
         message_number: u32,
     ) -> ServerResult<()> {
+        let transport = if let Some(transport) = self.transports.get(&transport_id) {
+            transport.clone()
+        } else {
+            return Err(ServerResultError::Internal(
+                ServerInternalError::TransportNotAttached,
+            ));
+        };
```
Comment (guidota) [inline rpc/src/server.rs:722]:
```suggestion
        let transport = self
            .transports
            .get(&transport_id)
            .ok_or(ServerResultError::Internal(
                ServerInternalError::TransportNotAttached,
            ))?
            .clone();

```

File: rpc/src/server.rs [hunk 5]
```diff
@@ -363,6 +368,26 @@ impl<Context: Send + Sync + 'static, T: Transport + ?Sized + 'static> RpcServer<
...
+                                    if tx_cloned
+                                        .send(TransportNotification::CloseTransport(id))
+                                        .is_err()
+                                    {
+                                        error!("> From a Transport > Error on while sending close notification for a transport")
+                                    }
+                                    return;
+                                } else {
+                                    transport.close().await;
+                                    return;
```
Comment (guidota) [inline rpc/src/server.rs:387]: Shouldn't we always call to `transport.close()`?

File: rpc/src/client.rs (hunk 1 above)
Comment (lauti7) [reply rpc/src/client.rs:88]: done!

File: rpc/src/transports/web_socket.rs (hunk 2 above)
Comment (lauti7) [reply rpc/src/transports/web_socket.rs:56]: removed!

File: rpc/src/transports/web_socket.rs (hunk 3 above)
Comment (lauti7) [reply rpc/src/transports/web_socket.rs:114]: done!

File: rpc/src/server.rs (hunk 4 above)
Comment (lauti7) [reply rpc/src/server.rs:722]: 💯

**Pull request merged**